design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,329 @@
+creating method signatures for as many input combinations as desired.
+- _Complex Aggregation Models_ - Today, KSQL only supports tracking a single variable for 
+aggregation and it must mirror the output type. To implement an `avg` UDAF, it is necessary to store 
+both a `count` and a `running_total`.
+
+## Scope
+
+All changes will be made for the new annotation-based UDF interfaces and will not be backported to
+the legacy implementation. The bullet points mentioned above are all in scope, but the following 
+will not be included in this KLIP:
+
+- UDTFs - multiple output operations are more complicated and will require a KLIP of their own
+- Generics/Varargs for UDAFs - the support for Generics/Varargs will only extend to UDFs.
+
+## Value
+
+Basic UDF operations that are supported in SQL languages will now be unblocked for KSQL. Some of
+these are outlined below:
+
+|      Improvement      |                                 Value                                  |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | Unblock `ARRAY_LENGTH`, `ARRAY_SLICE`, `ARRAY_TO_STRING`, etc...       |
+| Varargs support       | `CONCAT`, `MAX`, `MIN` etc...                                          |
+| Complex Aggregation   | `AVG`, `PERCENTILE`, etc...                                            |
+| Structured UDFs       | Arbitrary data handling without intermediary transforms                |
+
+## Public APIS
+
+Some public APIs will be changed, though all will be changed in a data backwards compatible fashion:
+
+|      Improvement      |                               API change                               |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | UDF interface will accept inferred generics (e.g. `List<T>`)           |
+| Varargs support       | UDF interface will accept `...` in the method declarations             |
+| Complex Aggregation   | UDAFs may include `VR export(A agg)` to convert aggregated to output   |
+| Structured UDFs       | UDF/UDAF interfaces will accept the Kafka Connect `Struct` type        |
+
+## Design
+
+Each of these improvements is relatively self-contained and can be worked on independently and in
+parallel. Below are detailed designs for each:
+
+### Structured UDFs
+
+After refactoring done in [#2411](https://github.com/confluentinc/ksql/pull/2411), the type coercion 
+ensures that the validation does not unnecessarily prevent usage of `Struct`. To complete support
+for sturcutred UDFS, we also need to have some mechanism to create the output schema. To address 
+this, we can resolve the schema as part of the UDF specification, requiring users to specify the 
+struct schema inline.
+
+```java
+@UdfSchema
+public static final Schema SCHEMA = SchemaBuilder.struct()...build();
+
+@Udf(schema="STRUCT<'VAL' VARCHAR, 'LENGTH' INT>")
```

Comment:
Minor point but probably good to clarify: are the single-quotes around the struct field names required? Are the field names case-sensitive?

Reply:
I will use the exact same syntax as `tableElement` in SqlBase.g4. I think you are correct in that the quotes are not required and it is not case sensitive (unless you quote, but that behavior is broken as of today)